src/thinblock.cpp

```diff
@@ -740,181 +737,207 @@ bool CXThinBlock::process(CNode *pfrom,
+    if (setHashesToRequest.size() > 0)
     {
-        // We have all the transactions now that are in this block: try to reassemble and process.
-        pfrom->thinBlockWaitingForTxns = -1;
-        int blockSize = pfrom->thinBlock.GetSerializeSize(SER_NETWORK, CBlock::CURRENT_VERSION);
-        LogPrint("thin", "Reassembled thin block for %s (%d bytes). Message was %d bytes, compression ratio %3.2f\n",
-            pfrom->thinBlock.GetHash().ToString(), blockSize, pfrom->nSizeThinBlock,
-            ((float)blockSize) / ((float)pfrom->nSizeThinBlock));
+        pfrom->thinBlockWaitingForTxns = setHashesToRequest.size();
+        CXRequestThinBlockTx thinBlockTx(header.GetHash(), setHashesToRequest);
+        pfrom->PushMessage(NetMsgType::GET_XBLOCKTX, thinBlockTx);
 
-        // Update run-time statistics of thin block bandwidth savings
-        thindata.UpdateInBound(pfrom->nSizeThinBlock, blockSize);
-        string ss = thindata.ToString();
-        LogPrint("thin", "thin block stats: %s\n", ss.c_str());
-        requester.Received(GetInv(), pfrom, pfrom->nSizeThinBlock);
+        thindata.UpdateInBoundReRequestedTx(pfrom->thinBlockWaitingForTxns);
+        return true;
+    }
+
+    // If there are still any missing transactions then we must clear out the thinblock data
+    // and re-request a full block (This should never happen because we just checked the various pools).
+    if (missingCount > 0)
+    {
+        // Since we can't process this thinblock then clear out the data from memory
+        thindata.ClearThinBlockData(pfrom);
 
-        PV.HandleBlockMessage(pfrom, strCommand, pfrom->thinBlock, GetInv());
+        std::vector<CInv> vGetData;
+        vGetData.push_back(CInv(MSG_BLOCK, header.GetHash()));
+        pfrom->PushMessage(NetMsgType::GETDATA, vGetData);
+        return error("Still missing transactions for xthinblock: re-requesting a full block");
     }
-    else if (pfrom->thinBlockWaitingForTxns > 0)
+
+    // We now have all the transactions now that are in this block
+    pfrom->thinBlockWaitingForTxns = -1;
+    int blockSize = pfrom->thinBlock.GetSerializeSize(SER_NETWORK, CBlock::CURRENT_VERSION);
+    LogPrint("thin", "Reassembled thin block for %s (%d bytes). Message was %d bytes, compression ratio %3.2f\n",
+        pfrom->thinBlock.GetHash().ToString(), blockSize, pfrom->nSizeThinBlock,
+        ((float)blockSize) / ((float)pfrom->nSizeThinBlock));
+
+    // Update run-time statistics of thin block bandwidth savings
+    thindata.UpdateInBound(pfrom->nSizeThinBlock, blockSize);
+    LogPrint("thin", "thin block stats: %s\n", thindata.ToString().c_str());
+    requester.Received(GetInv(), pfrom, pfrom->nSizeThinBlock);
+
+    // Process the full block
+    PV.HandleBlockMessage(pfrom, strCommand, pfrom->thinBlock, GetInv());
+
+    return true;
+}
+
+static bool ReconstructBlock(CNode *pfrom, const bool fXVal, int &missingCount, int &unnecessaryCount)
+{
```

Comment:
This is now a common function.  Please have it take a string indicating the source (thin or xthin) and put that in the logs, otherwise it is very hard to work out what is going on.

Also this set of changes has broken the statistics.  Frequently seen in my logs:

2017-05-21 00:44:38.485124 nGetXthinCount is nan

And this is no longer cumulative, only a single block:

2017-05-21 00:44:32.711818 thin block stats: 1 thin block has saved 978.27KB of bandwidth